### ISSUE TITLE
JDBC ERROR on Sqoop export Usage

### ISSUE THREAD

**Issue by jaykelin** [crate/crate#3984]
ENV:
centos\crate 0.55.2\crate-jdbc-0.13.0\sqoop-1.4.5
sqoop command like this:

> sqoop export --direct --batch --num-mappers 50 --connect jdbc:crate://10.1.113.172:4300 --driver io.crate.client.jdbc.CrateDriver --table dm_list_play_d_new --export-dir /datapath --fields-terminated-by '|' --null-non-string '' --null-string '' --input-null-non-string '\\N' --input-null-string '\\N' --columns 'some columns'

ERROR stack info：

> 16/08/17 17:50:23 INFO client.transport: [crate_client] failed to get node info for {#transport#-1}{10.1.113.172}{10.1.113.172:4300}, disconnecting...
> ReceiveTimeoutTransportException[[][10.1.113.172:4300][cluster:monitor/nodes/liveness] request_id [0] timed out after [5001ms]]
>         at io.crate.shade.org.elasticsearch.transport.TransportService$TimeoutHandler.run(TransportService.java:675)
>         at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
>         at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
>         at java.lang.Thread.run(Thread.java:744)
> 16/08/17 17:50:23 INFO client.transport: [crate_client] failed to get node info for {#transport#-1}{10.1.113.172}{10.1.113.172:4300}, disconnecting...
> TransportException[transport stopped, action: cluster:monitor/nodes/liveness]
>         at io.crate.shade.org.elasticsearch.transport.TransportService$2.run(TransportService.java:202)
>         at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
>         at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
>         at java.lang.Thread.run(Thread.java:744)
> 16/08/17 17:50:23 ERROR manager.SqlManager: Error executing statement: java.sql.SQLException: Connect to '10.1.113.172:4300' failed
> java.sql.SQLException: Connect to '10.1.113.172:4300' failed
>         at io.crate.client.jdbc.CrateConnection.connect(CrateConnection.java:60)
>         at io.crate.client.jdbc.CrateDriver.connect(CrateDriver.java:76)
>         at java.sql.DriverManager.getConnection(DriverManager.java:571)
>         at java.sql.DriverManager.getConnection(DriverManager.java:233)
>         at org.apache.sqoop.manager.SqlManager.makeConnection(SqlManager.java:875)
>         at org.apache.sqoop.manager.GenericJdbcManager.getConnection(GenericJdbcManager.java:52)
>         at org.apache.sqoop.manager.SqlManager.execute(SqlManager.java:736)
>         at org.apache.sqoop.manager.SqlManager.execute(SqlManager.java:759)
>         at org.apache.sqoop.manager.SqlManager.getColumnInfoForRawQuery(SqlManager.java:269)
>         at org.apache.sqoop.manager.SqlManager.getColumnTypesForRawQuery(SqlManager.java:240)
>         at org.apache.sqoop.manager.SqlManager.getColumnTypes(SqlManager.java:226)
>         at org.apache.sqoop.manager.ConnManager.getColumnTypes(ConnManager.java:295)
>         at org.apache.sqoop.orm.ClassWriter.getColumnTypes(ClassWriter.java:1773)
>         at org.apache.sqoop.orm.ClassWriter.generate(ClassWriter.java:1578)
>         at org.apache.sqoop.tool.CodeGenTool.generateORM(CodeGenTool.java:96)
>         at org.apache.sqoop.tool.ExportTool.exportTable(ExportTool.java:64)
>         at org.apache.sqoop.tool.ExportTool.run(ExportTool.java:100)
>         at org.apache.sqoop.Sqoop.run(Sqoop.java:143)
>         at org.apache.hadoop.util.ToolRunner.run(ToolRunner.java:70)
>         at org.apache.sqoop.Sqoop.runSqoop(Sqoop.java:179)
>         at org.apache.sqoop.Sqoop.runTool(Sqoop.java:218)
>         at org.apache.sqoop.Sqoop.runTool(Sqoop.java:227)
>         at org.apache.sqoop.Sqoop.main(Sqoop.java:236)
> Caused by: NoNodeAvailableException[None of the configured nodes are available: [{#transport#-1}{10.1.113.172}{10.1.113.172:4300}]]
>         at io.crate.shade.org.elasticsearch.client.transport.TransportClientNodesService.ensureNodesAreAvailable(TransportClientNodesService.java:290)
>         at io.crate.shade.org.elasticsearch.client.transport.TransportClientNodesService.execute(TransportClientNodesService.java:207)
>         at io.crate.client.InternalCrateClient.execute(InternalCrateClient.java:85)
>         at io.crate.client.InternalCrateClient.execute(InternalCrateClient.java:68)
>         at io.crate.client.InternalCrateClient.sql(InternalCrateClient.java:57)
>         at io.crate.client.CrateClient.sql(CrateClient.java:145)
>         at io.crate.client.jdbc.CrateStatement.execute(CrateStatement.java:65)
>         at io.crate.client.jdbc.CrateStatementBase.executeQuery(CrateStatementBase.java:37)
>         at io.crate.client.jdbc.CrateDatabaseMetaData.getDatabaseProductVersion(CrateDatabaseMetaData.java:79)
>         at io.crate.client.jdbc.CrateConnection.connect(CrateConnection.java:57)
>         ... 22 more
> 16/08/17 17:50:23 ERROR tool.ExportTool: Encountered IOException running export job: java.io.IOException: No columns to generate for ClassWriter

**Comment by jaykelin** [comment]
> [2016-08-17 21:03:22,808][WARN ][transport.netty          ] [crate172] exception caught on transport layer [[id: 0xd9740b6a, /10.1.113.172:5769 => /10.1.113.172:4300]], closing connection
> java.lang.IllegalStateException: Message not fully read (request) for requestId [0], action [cluster/nodes/info], readerIndex [39] vs expected [57]; resetting
>         at org.elasticsearch.transport.netty.MessageChannelHandler.messageReceived(MessageChannelHandler.java:120)
>         at org.jboss.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70)
>         at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
>         at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791)
>         at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:296)
>         at org.jboss.netty.handler.codec.frame.FrameDecoder.unfoldAndFireMessageReceived(FrameDecoder.java:462)
>         at org.jboss.netty.handler.codec.frame.FrameDecoder.callDecode(FrameDecoder.java:443)
>         at org.jboss.netty.handler.codec.frame.FrameDecoder.messageReceived(FrameDecoder.java:303)
>         at org.jboss.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70)
>         at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
>         at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791)
>         at org.elasticsearch.common.netty.OpenChannelsHandler.handleUpstream(OpenChannelsHandler.java:75)
>         at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
>         at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:559)
>         at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:268)
>         at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:255)
>         at org.jboss.netty.channel.socket.nio.NioWorker.read(NioWorker.java:88)
>         at org.jboss.netty.channel.socket.nio.AbstractNioWorker.process(AbstractNioWorker.java:108)
>         at org.jboss.netty.channel.socket.nio.AbstractNioSelector.run(AbstractNioSelector.java:337)
>         at org.jboss.netty.channel.socket.nio.AbstractNioWorker.run(AbstractNioWorker.java:89)
>         at org.jboss.netty.channel.socket.nio.NioWorker.run(NioWorker.java:178)
>         at org.jboss.netty.util.ThreadRenamingRunnable.run(ThreadRenamingRunnable.java:108)
>         at org.jboss.netty.util.internal.DeadLockProofWorker$1.run(DeadLockProofWorker.java:42)
>         at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
>         at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
>         at java.lang.Thread.run(Thread.java:745)

server log info

**Comment by mikethebeer** [comment]
@jaykelin 
```
Caused by: NoNodeAvailableException[None of the configured nodes are available: [{#transport#-1}{10.1.113.172}{10.1.113.172:4300}]]
```

are you sure that Crate is running and Port 4300 is not blocked by firewall?

**Comment by jaykelin** [comment]
@mikethebeer
i have already fixed this situation by adjust file crate.yml

> bootstrap.mlockall: true
> node.name: "crate175"
> gateway.expected_nodes: 3
> gateway.recover_after_nodes: 2
> discovery.zen.minimum_master_nodes: 2
> discovery.zen.ping:
>   multicast.enabled: false
>   unicast.hosts:
>     - crate172:4300
>     - crate175:4300
>     - crate176:4300
> es.api.enabled: true
> cluster.name: bi_crate
> cluster.graceful_stop.min_availability: full
> cluster.graceful_stop.reallocate: true
> cluster.graceful_stop.timeout: 10s
> cluster.graceful_stop.force: true
> node.master: true
> node.data: true
> stats.enabled: yes
> index.refresh_interval: 30s
> index.translog.sync_interval: 10s
> network.host: 10.1.113.175
> transport.tcp.port: 4300
> transport.tcp.connect_timeout: 10s
> indices.store.throttle.type: none
> indices.store.throttle.max_bytes_per_sec: 400mb
> indices.memory.index_buffer_size: 30%
> indices.memory.min_index_buffer_size: 256m
> indices.memory.min_shard_index_buffer_size: 28m
> discovery.zen.fd.ping_timeout: 120s
> discovery.zen.fd.ping_retries: 6
> discovery.zen.fd.ping_interval: 10s
> threadpool.index.type: fixed
> threadpool.search.type: fixed
> threadpool.get.type: fixed
> threadpool.bulk.type: fixed
> threadpool.refresh.type: scaling
> threadpool.index.queue_size: 2000
> threadpool.bulk.queue_size: 1000
> cluster.info.update.interval: 10s
> bulk.request_timeout: 10m
> udc.enabled: true
> udc.initial_delay: 1m
> udc.interval: 30m

Caused by: NoNodeAvailableException[None of the configured nodes are available: [{#transport#-1}{10.1.113.172}{10.1.113.172:4300}]]
To solve this exception，particulary I adjusted the following parameters
 > index.refresh_interval: 30s
> threadpool.index.queue_size: 2000
> threadpool.bulk.queue_size: 1000
> discovery.zen.fd.ping_timeout: 120s
> discovery.zen.fd.ping_retries: 6
> discovery.zen.fd.ping_interval: 10s

In addition, I also adjust the system parameters

> sysctl -w net.ipv4.tcp_keepalive_time=600
> sysctl -w net.ipv4.tcp_keepalive_probes=6
> sysctl -w net.ipv4.tcp_keepalive_intvl=60
